### PR TITLE
Increase resiliency of docker driver logging

### PR DESCRIPTION
**Pull request by dantoml**
- Adds a retry loop to streaming from the docker daemon
- Adds a respawn attempt when recovering tasks in case of docker_logger
failure

See commits for details about individual changes.

related to #5296

### CODE REVIEW

File: drivers/docker/driver.go [hunk 1]
```diff
@@ -138,14 +180,17 @@ func (d *Driver) RecoverTask(handle *drivers.TaskHandle) error {
...
+		d.logger.Warn("failed to reattach to docker logger process", "error", err)
 
-	dlogger, dloggerPluginClient, err := docklog.ReattachDockerLogger(reattach)
-	if err != nil {
-		return fmt.Errorf("failed to reattach to docker logger process")
+		dlogger, dloggerPluginClient, err = d.setupNewDockerLogger(container, handle.Config, time.Now())
+		if err != nil {
+			// TODO(dani): FIXME: Here we will leak the users container. Simply removing
+			//                    the container here will cause a user restart-stanza
+			//                    based cycle.
```
Comment (dantoml) [inline drivers/docker/driver.go]: https://github.com/hashicorp/nomad/pull/5332 is part of the fix for this

File: drivers/docker/docklog/docker_logger.go [hunk 2]
```diff
@@ -75,18 +80,39 @@ func (d *dockerLogger) Start(opts *StartOpts) error {
...
+			sinceTime = time.Now()
 
-	go func() { client.Logs(logOpts) }()
+			container, err := client.InspectContainer(opts.ContainerID)
+			if err != nil {
+				_, notFoundOk := err.(*docker.NoSuchContainer)
+				if !notFoundOk {
+					return
+				}
+			} else if container.State.Running != true {
```
[the remark this answers is not in the document]
Comment (dantoml) [reply drivers/docker/docklog/docker_logger.go]: Nope - The `container` here is a point in time snapshot from the docker api

[the remark this answers is not in the document]
Comment (schmichael) [reply drivers/docker/docklog/docker_logger.go]: Oh duh, from a few lines up. Sorry for the noise

**Comment by dantoml** [comment]
🐛 : Need to store the new plugin restore handle

File: drivers/docker/docklog/docker_logger.go [hunk 3]
```diff
@@ -75,18 +80,39 @@ func (d *dockerLogger) Start(opts *StartOpts) error {
...
+			}
+
+			sinceTime = time.Now()
 
-	go func() { client.Logs(logOpts) }()
+			container, err := client.InspectContainer(opts.ContainerID)
+			if err != nil {
+				_, notFoundOk := err.(*docker.NoSuchContainer)
+				if !notFoundOk {
+					return
```
[the remark this answers is not in the document]
Comment (dantoml) [reply drivers/docker/docklog/docker_logger.go]: This is fairly normal - When you stop/rm a container, the stream will close and then the container will 404

File: drivers/docker/docklog/docker_logger.go [hunk 4]
```diff
@@ -75,18 +80,42 @@ func (d *dockerLogger) Start(opts *StartOpts) error {
...
-		ErrorStream:  d.stderr,
-		Since:        0,
-		Follow:       true,
-		Stdout:       true,
-		Stderr:       true,
-	}
+	go func() {
+		sinceTime := time.Unix(opts.StartTime, 0)
+
+		for {
```
Comment (notnoop) [inline drivers/docker/docklog/docker_logger.go]: Should we add some backoff if we are seeing unexpected non-return-causing errors?

**Pull request merged**